### DISCUSSION TITLE
Cross Compile for beginners :-/

### DISCUSSION THREAD

**Discussion by robbieb43** [Barracuda09/SATPI#143] [Q&A] [answered]
Sorry for the lame nature of this request but I have gone round in a few circles and I think I need an experienced hand on my shoulder.

I have used satpi extensively on my pi based streaming servers for some years, but stuck with some elderly minisatip binaries on my enigma2 boxes.  I would like to migrate to satpi compiled for these environments (dm800 mips,   Vu se mips2 BCM7346B, + hopefully zgemma ARMv7 BCM7251S ).

My compilation environment is Ubuntu 20.02 based.  I have repeatedly installed the Broadcom tool chain as per the readme, but do not understand the bit about "get the sys root like headers and libraries for your device". My efforts so far get the predictable errors about missing files. I am not sure whether I need access to the source repository for my box(es) or my firmware variant, or whether these are binaries from my box.  If it is the rep, any clues as to where in the directory tree?

My skill level is: Linux, scripting and s/w development, etc - competent;  "Make" scripting - basic; gcc etc intermediate, cross-compilation basic but well versed with the concepts.

Any pointers or a pointer to the Faq that has clearly eluded me for some years would be very much gratefully received.

Many thanks

Rob

**Comment by Barracuda09** [comment, marked as answer]
Hi @robbieb43 

Thanks for you interest in SatPI.

Where did you "install" `https://github.com/Broadcom/stbgcc-8.3/releases`
I have 'installed' it in `~/vu/opt/toolchains/stbgcc-8.3-0.4/` so in my Home directory

Here is the wiki on how to build SatPI:
https://github.com/Barracuda09/SATPI/wiki/Build-SatPI

commands:
- `source env.sh` is for ARM based chips using the Broadcom Toolchain
- `source envMIPS.sh` is for MIPS based chips using the Broadcom Toolchain

 These scripts use Broadcom Toolchain in `~/vu/opt/toolchains/stbgcc-8.3-0.4`

_"get the sys root like headers and libraries for your device"_ this is not needed if you use Broadcom Toolchain

Kind regards
Marc

**Comment by robbieb43** [comment]
Thanks Marc - I think I need to start from scratch again with a clean VM as I cannot even answer your question reliably. Thanks for the FAQ and I think the bit about not needing the headers is the piece of key info that I needed.  I will have another go from scratch and report back.

thanks for a great project.

Rob

**Comment by Saentist** [comment]
If someone can modificate this builder to make SATPI
https://github.com/gorgone/s3_releases
most interesting if used as 
```
./s3 menu
```
this will solve lot of STB platforms building

**Comment by robbieb43** [comment]
@Barracuda09 Hi Marc - this worked.  With a clean VM I installed the toolchain in its default location (/opt/vu/toolchains etc) which for simplicity I then copied to my user home directory (~/vu/opt/toolchains etc) and it worked without further intervention.  So far I have created an executable for my zgemma ARM based box and its appears to be working, and a MIPS one, that on my VU clone and DM800 are asking for various dependencies. Sadly as they are a couple of thousand miles away I do not want to risk to much experimentation on them at the moment, so will need to fire-up a local MIPS box as test platform. The fact that they were able to report the dependency issue suggests a compatible architecture build though.  Many thanks  Rob.

**Comment by Barracuda09** [comment]
@robbieb43 

Be sure to open a different terminal for `source envMIPS.sh` and `source env.sh` before compiling and do a clean in between builds

**Reply by robbieb43** [reply]
Thanks Marc - good point.  Happy to report the ARM version successfully streamed some channels.

**Reply by Barracuda09** [reply]
What does this show for MIPS boxes  `cat /proc/cpuinfo`

**Reply by robbieb43** [reply]
VU clone:

system type             : BCM7346B2 STB platform
machine                 : Unknown
processor               : 0
cpu model               : Broadcom BMIPS5000 V1.1  FPU V0.1
BogoMIPS                : 864.25
cpu MHz                 : 1305.080
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 64
extra interrupt vector  : yes
hardware watchpoint     : no
isa                     : mips1 mips2 mips32r1
ASEs implemented        :
shadow register sets    : 1
kscratch registers      : 0
core                    : 0
VCED exceptions         : not available
VCEI exceptions         : not available

processor               : 1
cpu model               : Broadcom BMIPS5000 V1.1  FPU V0.1
BogoMIPS                : 655.36
cpu MHz                 : 1305.080
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 64
extra interrupt vector  : yes
hardware watchpoint     : no
isa                     : mips1 mips2 mips32r1
ASEs implemented        :
shadow register sets    : 1
kscratch registers      : 0
core                    : 0
VCED exceptions         : not available
VCEI exceptions         : not available

Dream Multimedia DM800

system type             : BCM97xxx Settop Platform
build target            : dm800
processor               : 0
cpu model               : Brcm3300 V0.0
cpu MHz                 : 294.91
BogoMIPS                : 294.91    ( udelay_val : 147456  HZ = 1000 )
wait instruction        : yes
microsecond timers      : yes
tlb_entries             : 32
extra interrupt vector  : yes
hardware watchpoint     : no
ASEs implemented        : mips16
VCED exceptions         : not available
VCEI exceptions         : not available
RAC setting             : I/D-RAC enabled
unaligned access        : 34558
rdhwr/brdhwr traps      : 0 / 0

The executable seemed to be recognised  as when I installed the first missing lib it moved on to ask for something else. I got nervous at this point about installing anything further as I will not be able to sort out any problems as I am remote. From the box.

**Reply by robbieb43** [reply]
The zgemma h7 (ARM) which is fully working is:

processor       : 0
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 54.00
Features        : half thumb fastmult vfp edsp thumbee neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm 
CPU implementer : 0x42
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0x00f
CPU revision    : 3

processor       : 1
model name      : ARMv7 Processor rev 3 (v7l)
BogoMIPS        : 54.00
Features        : half thumb fastmult vfp edsp thumbee neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm 
CPU implementer : 0x42
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0x00f
CPU revision    : 3

Hardware        : Broadcom STB (Flattened Device Tree)
Revision        : 0000
Serial          : 0000000000000000

**Comment by Barracuda09** [comment]
@robbieb43 

with this you can check which libs you need `objdump -p ./satpi | grep NEEDED`
Then you can put the libs in ex.`/usr/satpi` and execute bin with:

` LD_LIBRARY_PATH=/usr/satpi/ /usr/satpi/satpi`

**Comment by Barracuda09** [comment]
you can grab them from `vu/opt/toolchains/stbgcc-8.3-0.4/mips-unknown-linux-gnu/lib/`

**Reply by robbieb43** [reply]
Thanks Marc - I have now got a local test DM800 running with the same firmware. 

  NEEDED               librt.so.1
  NEEDED               libatomic.so.1
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libpthread.so.0
  NEEDED               libc.so.6

Your suggestion moved things forward but I now have hit another issue: it appears that the toolchain libs are not encoded in the right way, as the error message seems to imply:
"./libatomic.so.1: ELF file data encoding not little-endian".

I wonder if there is an option to produce a static binary that includes the libs?  My other option may be to try and search for them but I guess that my require a degree of luck.

Reverting to the software feed for my STB image, the version of libatomic seems to satisfy the above. I then installed libstdc++.so.6 but was not so lucky:

./satpi: /usr/lib/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by ./satpi)
./satpi: /usr/lib/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./satpi)

I will have to have a think.

Thanks again

Rob

**Reply by robbieb43** [reply]
I wonder if this suggests MIPSEL rather than MIPS - I will experiment.

**Reply by Barracuda09** [reply]
Yes you are correct it should most likely be MIPSEL. Just make a envMIPSEL.sh and point it to `vu/opt/toolchains/stbgcc-8.3-0.4/mipsel-unknown-linux-gnu/ `

**Reply by robbieb43** [reply]
Thanks - that seems to run without complaining. It is not sadly listening on the the specified TCP  port (though it is on 1900 and 1024 UDP).  Looks like we are getting there.  I will try producing a debug version tomorrow.

**Reply by robbieb43** [reply]
Hi Marc - I think I have hit a brick wall for now.  It is clearly running on the platform and creates the SatPi.xml file, populated with correct information about the FE etc.  I can only assume it is something to do with the linux version of my firmware.  Sadly new firmware versions are a little scarce but I will have a search.  Is there any way of getting more debug info out of the executable?

Thanks again Rob

**Reply by Barracuda09** [reply]
Can you access the webif? and copy-past the log?

**Comment by Barracuda09** [comment]
What is actually not working on the MIPSEL version?

**Reply by robbieb43** [reply]
It listens on the two UDP ports (1900 + 1024), but not on the TCP ports.  I have tried specifying ports for the http ports and rtsp but it does not show up (hence no browser log).  It creates the SatPi xml file with the details specified.  I have tried the --no-daemon option but there is no log output at the terminal.

**Reply by robbieb43** [reply]
sorry UDP 1029.
udp        0      0 0.0.0.0:1029            0.0.0.0:*                           10546/satpi
udp        0      0 0.0.0.0:1900            0.0.0.0:*                           10546/satpi

Linux kernel is old:

Linux version 2.6.18-7.4-dm800

**Reply by Barracuda09** [reply]
Yes that is rather old. Strange that there is no output with `--no-daemon` in the console. That suggest the libs are not linked with the correct (old) version. You could try to copy them from the box and try to link against them.

**Reply by robbieb43** [reply]
Perhaps the problem is a bit more basic:

I loaded a newer version of the firmware (openpli 7) although still is same kernel version. 

Using the LD_LIBRARY approach get the same result and noted this:

Mem: 95356K used, 50020K free, 0K shrd, 80K buff, 67948K cached
CPU:  61% usr  32% sys   0% nic   0% idle   0% io   0% irq   5% sirq
Load average: 0.98 0.72 0.55 3/48 3957

I wonder if it is simply a resources issue.  

However, exploring the feed I seemed to be able to satisfy the dependencies so trying again with a normal command line:

I get this:
4009  2512 root     S    33412  23%   0% ./satpi/satpi --no-daemon

and just the UDP ports unless I specify --no-ssdp (i.e. as before)


BTW on the VU+ clone I just get a segmentation fault, but since this is remote from me I did not want to explore further for fear of knocking it over.  

I suggest I close the issue but happy to volunteer to test future builds if anything comes to light or if others have more success.

Many thanks for all the support.

Rob

**Comment by Barracuda09** [comment]
Hi @robbieb43 

Thanks for your effort so far. Did you check this HowTo maybe this will work:
https://www.opena.tv/howtos/54527-satpi-satip-server-im-openatv-image.html

**Comment by robbieb43** [comment]
Many thanks for this.

It might provide a solution for my original problem as my remote DM800 is a genuine box and so I hopefully can get a version of OpenATV 6.3/4/5 for it.  That sadly will have to wait until the spring when I will be able to upgrade it.  In the meantime I might explore unpacking a firmware image to find out their feed url and see about getting hold of the executable.  If I make any advances, I will let you know.  

Thanks again for all the help.

Rob
